shell/platform/android/io/flutter/embedding/android/FlutterActivity.java

```diff
@@ -533,82 +533,132 @@ private void configureStatusBarForFullscreenFlutterExperience() {
   @Override
   protected void onStart() {
     super.onStart();
+    ensureAlive();
     lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_START);
     delegate.onStart();
   }
 
   @Override
   protected void onResume() {
     super.onResume();
+    ensureAlive();
     lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_RESUME);
     delegate.onResume();
   }
 
   @Override
   public void onPostResume() {
     super.onPostResume();
+    ensureAlive();
     delegate.onPostResume();
   }
 
   @Override
   protected void onPause() {
     super.onPause();
+    ensureAlive();
     delegate.onPause();
     lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_PAUSE);
   }
 
   @Override
   protected void onStop() {
     super.onStop();
-    delegate.onStop();
+    if (delegate != null) {
+      delegate.onStop();
+    } else {
+      Log.v(TAG, "FlutterActivity " + this + " onStop called after release.");
+    }
     lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_STOP);
   }
 
   @Override
   protected void onSaveInstanceState(Bundle outState) {
     super.onSaveInstanceState(outState);
-    delegate.onSaveInstanceState(outState);
+    if (delegate != null) {
+      delegate.onSaveInstanceState(outState);
+    } else {
+      Log.v(TAG, "FlutterActivity " + this + " onSaveInstanceState called after release.");
+    }
+  }
+
+  /**
+   * Irreversibly release this activity's control of the {@link FlutterEngine} and its
+   * subcomponents.
+   *
+   * <p>Calling will disconnect this activity's view from the Flutter renderer, disconnect this
+   * activity from plugins' {@link ActivityControlSurface}, and stop system channel messages from
+   * this activity.
+   *
+   * <p>After calling, this activity should be disposed immediately and not be re-used.
+   */
+  private void release() {
+    delegate.onDestroyView();
+    delegate.onDetach();
+    delegate.release();
+    delegate = null;
+  }
+
+  @Override
+  public void detachFromFlutterEngine() {
+    Log.v(
+        TAG,
+        "FlutterActivity "
+            + this
+            + " connection to the engine "
+            + getFlutterEngine()
+            + " evicted by another attaching activity");
+    release();
   }
 
   @Override
   protected void onDestroy() {
     super.onDestroy();
-    delegate.onDestroyView();
-    delegate.onDetach();
+    if (delegate != null) {
+      release();
+    } else {
+      Log.v(TAG, "FlutterActivity " + this + " onDestroy called after release.");
+    }
     lifecycle.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY);
   }
 
   @Override
   protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+    ensureAlive();
     delegate.onActivityResult(requestCode, resultCode, data);
   }
 
   @Override
   protected void onNewIntent(@NonNull Intent intent) {
     // TODO(mattcarroll): change G3 lint rule that forces us to call super
     super.onNewIntent(intent);
+    ensureAlive();
```

Comment:
I don't think this is needed since these methods would not normally be called by user code. They are entirely controlled by Android lifecycle itself and Android guarantees not calling these methods once activity is destroyed.

`ensureAlive` type thing made some sense in delegate because delegate could be used in many objects (including fragment and activity) it had to assert its own internal state.

Reply:
makes sense. Thanks